src/Confluent.Kafka/ProducerBuilder.cs

```diff
@@ -0,0 +1,292 @@
+
+        /// <summary>
+        ///     The configured async key serializer.
+        /// </summary>
+        internal protected IAsyncSerializer<TKey> AsyncKeySerializer { get; set; }
+
+        /// <summary>
+        ///     The configured async value serializer.
+        /// </summary>
+        internal protected IAsyncSerializer<TValue> AsyncValueSerializer { get; set; }
+
+        internal ProducerBase.Config ConstructBaseConfig(Producer<TKey, TValue> producer)
+        {
+            return new ProducerBase.Config
+            {
+                config = Config,
+                errorHandler = this.ErrorHandler == null
+                    ? default(Action<Error>) // using default(...) rather than null (== default(...)) so types can be inferred.
+                    : error => this.ErrorHandler(producer, error),
+                logHandler = this.LogHandler == null
+                    ? default(Action<LogMessage>)
+                    : logMessage => this.LogHandler(producer, logMessage),
+                statisticsHandler = this.StatisticsHandler == null
+                    ? default(Action<string>)
+                    : stats => this.StatisticsHandler(producer, stats)
+            };
+        }
+
+        /// <summary>
+        ///     A collection of librdkafka configuration parameters 
+        ///     (refer to https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md)
+        ///     and parameters specific to this client (refer to: 
+        ///     <see cref="Confluent.Kafka.ConfigPropertyNames" />).
+        ///     At a minimum, 'bootstrap.servers' must be specified.
+        /// </summary>
+        public ProducerBuilder(IEnumerable<KeyValuePair<string, string>> config)
+        {
+            this.Config = config;
+        }
+
+        /// <summary>
+        ///     Set the handler to call on librdkafka statistics events. Statistics are provided as a JSON formatted string as defined here:
+        ///     https://github.com/edenhill/librdkafka/wiki/Statistics
+        /// </summary>
+        /// <remarks>
+        ///     You can enable statistics and set the statistics interval
+        ///     using the statistics.interval.ms configuration parameter
+        ///     (disabled by default).
+        ///
+        ///     Executes on the poll thread (by default, a background thread managed by
+        ///     the producer).
+        /// </remarks>
+        public ProducerBuilder<TKey, TValue> SetStatisticsHandler(Action<Producer<TKey, TValue>, string> statisticsHandler)
+        {
+            this.StatisticsHandler = statisticsHandler;
+            return this;
+        }
+
+        /// <summary>
+        ///     Set the handler to call on error events e.g. connection failures or all
+        ///     brokers down. Note that the client will try to automatically recover from
+        ///     errors that are not marked as fatal - such errors should be interpreted
```

Comment:
"such errors" is ambigious, does it refer to fatal or non-fatal errors? Better clarify.